pkg/openstack/instances.go

```diff
@@ -677,26 +721,91 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
 	return addrs, nil
 }
 
-func getAddressesByName(client *gophercloud.ServiceClient, name types.NodeName, networkingOpts NetworkingOpts) ([]v1.NodeAddress, error) {
-	srv, err := getServerByName(client, name)
+func getAddressesByName(compute *gophercloud.ServiceClient, name types.NodeName, networkingOpts NetworkingOpts, network *gophercloud.ServiceClient) ([]v1.NodeAddress, error) {
+	srv, err := getServerByName(compute, name)
 	if err != nil {
 		return nil, err
 	}
 
-	interfaces, err := getAttachedInterfacesByID(client, srv.ID)
+	interfaces, err := getAttachedInterfacesByID(compute, srv.ID, network)
 	if err != nil {
 		return nil, err
 	}
 
 	return nodeAddresses(&srv.Server, interfaces, networkingOpts)
 }
 
+// getSubInterfaces
+func getSubInterfaces(network *gophercloud.ServiceClient) ([]attachinterfaces.Interface, error) {
+	var interfaces []attachinterfaces.Interface
+
+	// Check if trunk ports are attached
+	listOpts := trunks.ListOpts{}
+	allPages, err := trunks.List(network, listOpts).AllPages()
+	if err != nil {
+		klog.Errorf("Failed to list trunks: %v", err)
+		return interfaces, err
+	}
+	allTrunks, err := trunks.ExtractTrunks(allPages)
+	if err != nil {
+		klog.Errorf("Failed to extract trunks: %v", err)
+		return interfaces, err
+	}
+
+	// Get subports attached to the trunk
+	var subports []trunks.Subport
+	for _, trunk := range allTrunks {
+		for _, iface := range interfaces {
+			if iface.PortID == trunk.PortID {
+				s, err := trunks.GetSubports(network, trunk.ID).Extract()
+				if err != nil {
+					klog.Errorf("Failed to get subports for trunk %s: %v", trunk.ID, err)
+					return interfaces, err
+				}
+				subports = append(subports, s...)
+			}
+		}
+	}
+	klog.V(5).Infof("subports %v", subports)
+
+	// Arrange subports as for directly attached ports
+	for _, sport := range subports {
+		p, err := ports.Get(network, sport.PortID).Extract()
+		if err != nil {
+			klog.Errorf("Failed to get port info for subport %s: %v", sport.PortID, err)
+			return interfaces, err
+		}
+		n, err := networks.Get(network, p.NetworkID).Extract()
```

Comment:
A network query per subport doesn't seem like a good idea.

Reply:
Need to get the network name for filtering on internal-network-name later on.
https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/openstack-cloud-controller-manager/using-openstack-cloud-controller-manager.md#networking

Reply:
Point taken, we will keep this mind when working on PR update.

Reply:
We could probably list networks and then use that list to find needed info. It has downsides too.